[chip] Migrate component [Avatar] Support children, change default colors

mbrookes opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! -->

- [x] PR has tests / docs demo, and is linted.
- [x] Commit and PR titles begin with [ComponentName], and are in imperative form: "[Component] Fix leaky abstraction".
- [x] Description explains the issue / use-case resolved, and auto-closes the related issue(s) (http://tr.im/vFqem).

Issues: 
 - ~I've changes the default colours for `Avatar`, but directly, not with palette - I could use some guidance on what should or shouldn't be added to palette.~
- ~Not sure I'm applying styles to sub-components in the most optimal way - I've had to add a bunch of `SubComponentClassName` props. Not sure that's such a bad thing in and of itself, but again would welcome guidance.~
- ~I should probably separate the `Avatar` and `Chip` changes, but one was dependant on the other, so are lumped together for now.~
- ~I had some difficulty finding the right selector one of the tests: [Chip.spec.js:64](https://github.com/callemall/material-ui/pull/5852/files#diff-047ff436cab126e015d1cebb8c293ed3R64)~
- Chip probably needs more test cases. Suggestions welcome!

~Oh, and to agree on `children` vs `icon` prop (or both) for `Avatar` - I've added `children` so we can compare, but I'm wide open on it.~

mbrookes committed:
[Chip] Migrate component
[Avatar] Support children, change default colors

mbrookes commented:
circleci is baarfing on `npm run test:coverage` but it's fine locally...

mbrookes commented:
Looks like whatever was up with the coverage test got fixed and the tests automatically re-run, but it was still failing on the visual regression test due to my changes to Avatar, meaning I was premature in removing  "PR: Needs Revision". I've updated the Avatar baseline images for now.

oliviertassinari commented on src/Avatar/Avatar.js:
```diff
@@ -77,6 +94,10 @@ Avatar.propTypes = {
   /**
+   * Used to render icon and text elements inside the Avatar.
+   */
+  children: PropTypes.node,
```
What was the issue with the icon property? I mean, we could have simply renamed it children. Why keeping both or do we need a different logic?

mbrookes replied:
Not keeping both, I just didn't want to barge in and change the API without discussing it first, so it left both options in place until everyone was agreed. I'll strip it back to just `children`.

oliviertassinari commented on docs/site/src/demos/chips/Chips.js:
```diff
@@ -0,0 +1,56 @@
+    margin: 6,
+  },
+  svgIcon: {
+    color: '#444',
```
Why not using a `grey` variation?

oliviertassinari commented on docs/site/src/demos/chips/Chips.js:
```diff
@@ -0,0 +1,56 @@
+
+const styleSheet = createStyleSheet('Chips', () => ({
+  chip: {
+    margin: 6,
```
the smallest unity of margin / gutter is `8dp`. Can we use 8 here?

mbrookes replied:
With a margin of 6px the separation of the icons is 12px, is that okay?

oliviertassinari replied:
I think that it would be good to have a theme property for that. We could use it on the `Layout` component.

oliviertassinari reviewed (commented):
That looks really great overall 😄 .

mbrookes commented on src/Chip/Chip.spec.js:
```diff
@@ -0,0 +1,76 @@
+
+    it('should render a div containing an Avatar div, span and svg', () => {
+      assert.strictEqual(wrapper.is('div'), true, 'should be a div');
+      assert.strictEqual(wrapper.childAt(0).prop('component'), 'div', 'should be a div');
```
I was struggling to find the right target for the test, so arrived at that syntax blindly, so more than happy to change it if this is the right thing to test for...

oliviertassinari replied:
That looks good to me 👍 .

mbrookes commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+  if (childCount > 1) {
+    children = React.Children.toArray(children);
+
+    if (React.isValidElement(children[0]) && children[0].type.muiName === 'Avatar') {
```
I did catch one comment about `muiName`, but not the full discussion. I may be mistaken, but I don't think we save an iteration by moving from children to an `avatar` prop, as still you have to check it's is a valid element, and clone it to apply styling.

oliviertassinari replied:
Oh right, we are only checking the first children. In that case, it's more a matter of API consistency.
I would rather depend on a named `avatar` property than a less obvious `muiName` checking.

mbrookes replied:
I missed this comment - I'll give it a go tomorrow.

mbrookes commented on src/svg-icons/navigation/cancel.js:
```diff
@@ -0,0 +1,19 @@
+  </SvgIcon>
+);
+const NavigationCancel = pure(NavigationCancelIcon);
+NavigationCancel.displayName = 'NavigationCancel';
```
Any idea why we had it then? Isn't it used in React dev tools?

oliviertassinari replied:
the `pure` Higher order component is already providing enough information. I think that we can save that line.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+  const avatarClassName = classNames(classes.avatar);
+  const avatarIconClassName = classNames(classes.avatarIcon);
+  const labelClassName = classNames(classes.label, labelClassNameProp);
+  const deleteIconClassName = classNames({ [classes.deleteIcon]: onRequestDelete }, deleteIconClassNameProp);
```
Might worse only computing that variable only when `onRequestDelete` is truthy.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+  const classes = context.styleManager.render(styleSheet);
+  const className = classNames(classes.root, { [classes.clickable]: onClick }, classNameProp);
+  const avatarClassName = classNames(classes.avatar);
+  const avatarIconClassName = classNames(classes.avatarIcon);
```
No need to call `classNames` with a single argument.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+
+  const classes = context.styleManager.render(styleSheet);
+  const className = classNames(classes.root, { [classes.clickable]: onClick }, classNameProp);
+  const avatarClassName = classNames(classes.avatar);
```
No need to call `classNames` with a single argument.

mbrookes replied:
I think I need to read the jss docs. :) 

Is `classNames` just concatenating  them?

oliviertassinari replied:
classNames is an helper. That has nothing to do with jss. https://github.com/JedWatson/classnames.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+    },
+    avatarIcon: {
+      width: 19.2, // Added for jss avatar
+      height: 19.2, // Added for jss avatar
```
Don't we have layouting issues with non-integer?

mbrookes replied:
Not sure - in `master` this is a dynamic style in Avatar, where the icon is 0.6 * avatar `size`. Since Avatar no longer supports this, we have to calculate it manually (32 * 0.6 = 19.2). 

Happy to round it down if you think it could cause a problem.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+      color: palette.getContrastText(palette.grey[300]),
+      backgroundColor: palette.grey[300],
+      boxShadow: shadows[0],
+      borderRadius: 16,
```
Would using `50%` be better? Otherwise, I would have explicitly `height / 2`.

mbrookes replied:
Yes, 50% is better.

mbrookes replied:
I added `const height = 32` and `height: height`, `borderRadius: height / 2` if that's what you meant.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+      border: 0,
+      color: palette.getContrastText(palette.grey[300]),
+      backgroundColor: palette.grey[300],
+      boxShadow: shadows[0],
```
What's the need for `boxShadow: 'none'`?

mbrookes replied:
I think it was the default for a dynamic style, so is no longer needed. Removed.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+      justifyContent: 'center',
+      height: 32,
+      outline: 'none',
+      border: 0,
```
It's applied on a `div` do we need that property?

mbrookes replied:
Ditto.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,168 @@
+      alignItems: 'center',
+      justifyContent: 'center',
+      height: 32,
+      outline: 'none',
```
It's applied on a `div` do we need that property?

mbrookes replied:
That was carried over from `master`, but can't imagine why it was needed. Removed.

oliviertassinari commented on src/Avatar/Avatar.js:
```diff
@@ -110,3 +135,5 @@ Avatar.defaultProps = {
   styleManager: PropTypes.object.isRequired,
 };
+
+Avatar.muiName = 'Avatar';
```
We could remove that line with the `avatar` api.

mbrookes replied:
At the moment it's used in Chip to determine the child type - I'll leave it until the Chip API is solidified.

oliviertassinari commented on src/Avatar/Avatar.js:
```diff
@@ -14,13 +14,17 @@ export const styleSheet = createStyleSheet('Avatar', (theme) => {
     defaultColor: {
-      color: palette.getContrastText(palette.text.disabled),
-      background: palette.text.disabled,
+      color: 'white',
```
Could be using
```jsx
import { white } from './colors';
```

mbrookes replied:
Is it really worth importing 'white'? 😄

mbrookes commented:
@oliviertassinari I've switched to your preferred API. Take a look!

In doing so, I also removed the `iconClassName` prop that I previously added to `Avatar`. 

This makes Avatar simpler, and removes an otherwise redundant prop from Avatar (users can style the Avatar icon directly), but it does make the Chip code more complex. I'm open to your suggestion on what's preferable here.

oliviertassinari commented on src/Avatar/Avatar.js:
```diff
@@ -14,13 +15,18 @@ export const styleSheet = createStyleSheet('Avatar', (theme) => {
+      color: palette.background.default,
+      backgroundColor: emphasize(palette.background.default, 0.26),
+    },
+    icon: {
```
What's the need for this style? It seems to be working fine with font and svg icons.
That would allow us to avoid the cloneElement.

mbrookes replied:
> That would allow us to avoid the cloneElement.

It would, but I had to put it back for childrenClassName. 

I agree the styles are redundant though - they're an artefact from Avatar on `master` where these were dynamic based on the `size` prop.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,172 @@
+    if (avatarChildren) {
+      // If Avatar has an icon, style it.
+      if (React.isValidElement(avatarChildren)) {
+        const avatarIconClassName = classNames(classes.avatarIcon, avatarChildren.props.className);
```
We are accessing `avatarProp.props.children.props.className`, dear lord, that sounds really wrong.
What about moving the responsibility by exposing a `childrenClassName` property to the `<Avatar />`?

mbrookes replied:
That's how I did it in a previous commit  (well, `iconClassName`) - see my comment above:

> In doing so, I also removed the iconClassName prop that I previously added to Avatar.

This makes Avatar simpler, and removes an otherwise redundant prop from Avatar (users can style the Avatar icon directly), but it does make the Chip code more complex. I'm open to your suggestion on what's preferable here.

I think making Avatar `childrenClassName` hidden in the docs is probably the best compromise.

oliviertassinari reviewed (changes requested):
Thanks for changing the API 👍 .
Think we are close to get it merged.

mbrookes commented on test/regressions/site/src/tests/Chip/DeletableAvatarChip.js:
```diff
@@ -0,0 +1,16 @@
+    <Chip
+      avatar={<Avatar>MB</Avatar>}
+      label="SvgIcon Chip"
+      onClick={() => {}}
```
Right, no - copy pasted from the docs demo.  Although that makes me curious - is it possible to regression test the hover / focus / active styles?

mbrookes commented on src/svg-icons/cancel.js:
```diff
@@ -0,0 +1,18 @@
+  <SvgIcon {...props}>
+    <path
+      d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41
+      8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
```
> Is a breaking line affeIs a breaking line affecting the path? 

No it isn't, and I had manually fixed the linting errors in about half the icons which included the line-length rule, then decided instead to add `/* eslint-disable */ (I guess I could have added  `src/svg-icons to `.eslintignore` instead) and returned the icons to their generated form. Looks like I missed this one.

mbrookes commented:
@oliviertassinari Okay, I've pushed those updates. My replies to your review comments may be hidden now.

A few outstanding things: 
 - Keyboard focus isn't working and I'm not sure why. (It's working on `master`).
 - Would it make sense for a click on the delete icon to only emphasize the delete icon, and not the whole chip? If so, how would I implement that that? `event.stopPropagation`?
 - Should I disable the Avatar image regression test? A test that always fails and as a result means the rest of the tests don't run seems worse than no test!

oliviertassinari commented:
> Keyboard focus isn't working and I'm not sure why. (It's working on master).

I don't think that we need that feature at the `Chip` component level. For instance, look at the angular material examples: https://material.angularjs.org/latest/demo/chips.

> Would it make sense for a click on the delete icon to only emphasize the delete icon

Good question, looking at the other implementation, I would answer no.
They tend to consider the chip as a single unit.
- Angular: ![capture d ecran 2017-01-03 a 21 06 52](https://cloud.githubusercontent.com/assets/3165635/21621280/9989774e-d1f8-11e6-9e21-8651b891ce02.png)
- Materializecss:  ![capture d ecran 2017-01-03 a 21 07 31](https://cloud.githubusercontent.com/assets/3165635/21621299/acceafb8-d1f8-11e6-9bac-1dcaf44a1489.png)

> Should I disable the Avatar image regression test?

No, it's not because we broke a test that we should delete it. Fixing it would be much better.
![capture d ecran 2017-01-03 a 21 03 23](https://cloud.githubusercontent.com/assets/3165635/21621220/4b7636f0-d1f8-11e6-98eb-bc5f9f97634c.png)
> `http://0.0.0.0:3333/#/ListItem/AvatarListItem`

You can reproduce that on the list demo.

I think that we are close 🚢 .

mbrookes commented on src/Avatar/Avatar.js:
```diff
@@ -44,18 +47,23 @@ export default function Avatar(props, context) {
-    children = icon;
+  if (childrenProp) {
+    // If it's an icon, style it.
+    if (React.isValidElement(childrenProp)) {
```
Thanks for your patience! - It seems like my coding ability has regressed 2 years in 6 months, :worried:

mbrookes replied:
Done.

mbrookes commented:
Going to pick up these loose ends tomorrow. Thanks so much for putting up with me!

mbrookes commented on src/Avatar/Avatar.js:
```diff
@@ -110,3 +126,4 @@ Avatar.defaultProps = {
 Avatar.contextTypes = {
   styleManager: PropTypes.object.isRequired,
 };
+
```
+1

mbrookes commented on src/Avatar/Avatar.spec.js:
```diff
@@ -5,6 +5,7 @@ import React from 'react';
 import Avatar, { styleSheet } from './Avatar';
+import FaceIcon from '../svg-icons/face';
 
 describe('<Avatar />', () => {
```
Done.

mbrookes commented on src/Chip/Chip.js:
````diff
@@ -0,0 +1,164 @@
+ *
+ * ```jsx
+ * <Chip>
+ *   <Avatar>
````
Yes.

mbrookes commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,164 @@
+  let deleteIcon = null;
+  if (onRequestDelete) {
+    deleteIconClassName = classNames(classes.deleteIcon, deleteIconClassNameProp);
+    deleteIcon =
```
👍

oliviertassinari commented:
> Thanks so much for putting up with me!

No, thanks for taking care of that component!

mbrookes commented:
> > Keyboard focus isn't working and I'm not sure why. (It's working on master).

>I don't think that we need that feature at the `Chip` component level. For instance, look at the angular material examples: https://material.angularjs.org/latest/demo/chips.

I wouldn't read too much in to what material.angularjs.org are doing - it's a pretty ugly implementation.

Are you saying we shouldn't apply any focus style, and let the parent component watch for focus and apply focus styles to chip? That doesn't seem right to me.

> > Would it make sense for a click on the delete icon to only emphasize the delete icon

> Good question, looking at the other implementation, I would answer no. They tend to consider the chip as a single unit.

Again, I wouldn't ready too much into that - they don't apply hover or active styes to the chip *or* close button, so there's no difference anyway. react-toolbox applies hover style to the close button, but nothing else.

That reminds me though - what's the preferred prop name for the close button? it's `onRequestDelete` at the moment, but on `onCloseClick` / `onDeleteClick` or `onClickClose` / `onClickDelete`to match `onClick` are possible candidates.

> > Should I disable the Avatar image regression test?

> No, it's not because we broke a test that we should delete it. Fixing it would be much better.

I didn't say delete, I said disable. :smile: I'm talking about the regression test that you said has a race-condition. If it can be fixed - go for it!

I fixed the docs demo issue you mentioned, which was simply because I had not updated it  to follow the revised API. (Incidentally, react-toolbox is using the same composed API, and similar code to that we just replaced with props...)

mbrookes commented:
Regarding focus - I'd forgotten that in `master` Chip is using `EnhancedButton` which handles the keyboard focus etc.

`ButtonBase` didn't exist (I don't think) when I started porting Chip to `jss-themes-experiment`, so this just uses a vanilla `div`. What to do? We need keyboard control for accessibility.

mbrookes commented:
I realised the reason keyboard focus wasn't working is that `Chip` on master uses `EnhancedButton` which takes care of it, along with default keyboard event handling. I don't think `ButtonBase` existed when I originally started porting Chip to `jss-theme-experiment`.

I have started to implement the relevant functionality in Chip, but I'm wonder if I should instead use `ButtonBase`. It's a lot of code to drag along for a few functions, but using it is more DRY than reimplementing existing keyboard handing in Chip...

oliviertassinari commented:
> it's a pretty ugly implementation.

Well, at least, it's kind of do the job. A co-worker started implementing that component at the office with our own style guild. I will see what we can learn from that.

> Are you saying we shouldn't apply any focus style, and let the parent component watch for focus

For the use case I was thinking of, chips wouldn't be keyboard focusable. The parent component would manually apply the focus state depending on it's own state.

> Should we channel all events through onChange?

I'm not convinced. I don't think that some variations are going to hurt. Users will still have to look at the documentation to know how to use it correctly. Either it's the right callback property name or how to filter events.

> We're back to the failing ImageAvatar

Ewww, we gonna need to address that flaky test.

> I'm wonder if I should instead use ButtonBase

I think that I would definitely use it if I needed the ripple effect or having a `button` rendered in the DOM.

mbrookes commented:
>>it's a pretty ugly implementation.
>Well, at least, it's kind of do the job. 

True - they have a `Chips` component. Our chip is pretty useless without one! 😅 

By "ugly" I meant the visual appearance of the chip - I notice they have have hover and click styling on their todo-list.

> A co-worker started implementing that component at the office with our product style guides. I will see what we can learn from that.

👍 

> For the use case I was thinking of, chips wouldn't be keyboard focusable. The parent component would manually apply the focus state depending on it's own state.

I know they're not directly comparable, but thinking of Chip as a subcomponent for some kind of input, I'm looking at `Tab` and `BottomNavigationButton` as analogs. They're both keyboard focusable without dependance on their parent component - albeit handled by `ButtonBase`.

If you're happy for me to leave it as is for now, we can revisit this in the context of a `ChipInput` / `ChipAutoComplete` if anyone tackles that.

> I think that I would definitely use the ButtonBase if I needed the ripple effect or having a button rendered in the DOM.

It doesn't need to ripple, or focus ripple. It should probably be a button though as it's an interactive element that can be clicked etc. That would fix the `jsx-a11y/no-static-element-interactions` error. 

I don't think we need to use `ButtonBase just for that though.

oliviertassinari commented on docs/site/src/demos/chips/ChipsArray.js:
```diff
@@ -0,0 +1,73 @@
+
+    const chipData = this.state.chipData;
+    const chipToDelete = chipData.map((chip) => chip.key).indexOf(key);
+    chipData.splice(chipToDelete, 1);
```
I'm not sure it's a great example. `splice` mutate the array. It would be good to clone `chipData` first 😁 .
```js
const chipData = [...this.state.chipData];
```

mbrookes replied:
Line 47? (Or am I misunderstanding?)

oliviertassinari replied:
Yes, indeed

mbrookes replied:
Right, right. I'm surprised React doesn't scream about that!

oliviertassinari commented on docs/site/src/demos/chips/Chips.js:
```diff
@@ -0,0 +1,65 @@
+  const classes = context.styleManager.render(styleSheet);
+  return (
+    <div className={classes.row}>
+      <Chip
```
This element is focusable but don't have `:focus` state.
Is that expected?
![capture d ecran 2017-01-10 a 23 53 46](https://cloud.githubusercontent.com/assets/3165635/21828303/101d16b2-d790-11e6-9e51-88a1275c9609.png)

oliviertassinari commented on docs/api/Chip/Chip.md:
```diff
@@ -0,0 +1,24 @@
+| onClick | function |  | Callback function fired when the `Chip` element is clicked.<br><br>**Signature:**<br>`function(event: object) => void`<br>*event:* TouchTap event targeting the element. |
+| onRequestDelete | function |  | Callback function fired when the delete icon is clicked. If set, the delete icon will be shown.<br><br>**Signature:**<br>`function(event: object) => void`<br>*event:* `touchTap` event targeting the element. |
+
+Other properties (not documented) are applied to the root element.
```
We might need to rebase on `next` as I think we have improved that wording thanks to you 👍

oliviertassinari commented on docs/site/src/demos/lists/FolderList.js:
```diff
@@ -23,11 +23,11 @@ function FolderList(props, context) {
         </ListItem>
         <ListItem button>
-          <Avatar icon={<span className="material-icons">folder</span>} />
+          <Avatar ><span className="material-icons">folder</span></Avatar>
```
```jsx
<Avatar><span className="material-icons">folder</span></Avatar>
```

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,211 @@
+
+  let chipRef;
+
+  const handleDeleteIconClick = (event) => {
```
I have never seen that pattern before. Interesting.
We could also share the function definition between renders by accepting a `props` argument.

mbrookes replied:
> We could also share the function definition between renders by accepting a props argument.

Sorry, could you explain please?

oliviertassinari replied:
If we change the signature to the following, I think that we would be able to allocate the function once.
```jsx
const handleDeleteIconClick = (event, props) => {
```

mbrookes replied:
Sorry you're going to have to spell it out. 😊 

Why pass `props` when we're not using it, and how does that affect allocation?

oliviertassinari replied:
Right, my bad, that's not going to change anything.
Using a class on the other hand could help.

mbrookes replied:
For `Chip`? I

t's a trade-off I guess. I wonder how much difference it would make, as instead we're dragging lifecycle methods along for the ride.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,211 @@
+  let deleteIcon = null;
+  if (onRequestDelete) {
+    deleteIconClassName = classNames(classes.deleteIcon, deleteIconClassNameProp);
+    deleteIcon = (<DeleteIcon className={deleteIconClassName} onClick={handleDeleteIconClick} />);
```
I don't think we need the parentheses `()`.

oliviertassinari commented on src/Chip/Chip.js:
```diff
@@ -0,0 +1,211 @@
+    <button
+      className={className}
+      onClick={onClick}
+      tabIndex={0}
```
I'm wondering if it shouldn't be `-1` when the element don't have `onClick` or `onRequestClose` callbacks.

mbrookes replied:
Yes, that should fix the focus issue you [mentioned](https://github.com/callemall/material-ui/pull/5852#discussion_r95476646).

mbrookes replied:
It fixed the issue, but now I'm wondering if there should no `tabIndex` at all when the element don't have `onClick` or `onRequestClose` callbacks?

oliviertassinari replied:
I'm wondering, isn't `tabIndex={0}` doing nothing on an enable button. In that case yes, it could be undefined.

oliviertassinari reviewed (commented):
I left some minor comments. That's some great stuff 🎉 .

mbrookes commented:
@oliviertassinari I had a couple of questions regarding your line comments (in case the revisions have hidden any of them).

oliviertassinari commented on docs/site/src/demos/chips/ChipsArray.js:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    const chipData = this.state.chipData;
+    const chipToDelete = chipData.map((chip) => chip.key).indexOf(key);
```
I think that it would be better like this, only reading memory and not allocating anything.
```jsx
const chipToDelete = chipData.indexOf(chipData.find((chip) => chip.key === key)));
```

oliviertassinari commented on docs/site/src/demos/lists/FolderList.js:
```diff
@@ -24,11 +24,11 @@ function FolderList(props, context) {
         </ListItem>
         <ListItem button>
-          <Avatar icon={<FolderIcon />} />
+          <Avatar> <FolderIcon /></Avatar>
```
```js
<Avatar><FolderIcon /></Avatar>
```

mbrookes replied:
Merge conflict foobar. Fixed.

oliviertassinari commented on docs/site/src/demos/chips/ChipsArray.js:
```diff
@@ -0,0 +1,73 @@
+      return;
+    }
+
+    const chipData = Object.assign([], this.state.chipData);
```
That's probably the only the only reference to `Object.assign` in the code base. I would rather use the destructuration.

mbrookes replied:
Oh, interesting. That sounds like a much better idea.

(I did wonder where all the `Object.assign`s were in `next` given how heavily we used it on `master` 😄 )
